cpp/src/rolling/detail/range_utils.cuh

```diff
@@ -0,0 +1,868 @@
+ * @tparam WindowTag The tag indicating the type of window being computed
+ * @tparam Order The sort order of the orderby column defining the window.
+ */
+template <direction Direction, window_tag WindowTag, cudf::order Order>
+struct range_window_clamper {
+  /**
+   * @brief Functor to compute distance from a given row to the edge
+   * of the window.
+   *
+   * @tparam Grouping Object defining how the orderby column is
+   * grouped.
+   * @tparam OrderbyT Type of elements in the orderby columns.
+   * @tparam DeltaT Type of the elements in the scalar delta (returned
+   * by scalar.data()).
+   * @param groups The grouping object.
+   * @param row_delta Pointer to row delta on device.
+   * @param begin Iterator to begin of orderby column on device.
+   * @param end Iterator to end of orderby column on device.
+   *
+   * @note Let `x` be the value of the current row and `delta` the provided
+   * row delta then for bounded windows the endpoints are computed as follows.
+   *
+   *           | ASCENDING | DESCENDING
+   * ----------+-----------+-----------
+   * PRECEDING | x - delta | x + delta
+   * FOLLOWING | x + delta | x - delta
+   *
+   * See `saturating_sub` and `saturating_add` for details of the implementation of
+   * saturating addition/subtraction.
+   */
+  template <typename Grouping, typename OrderbyT, typename DeltaT>
+  struct distance_kernel {
+    distance_kernel(Grouping groups,
+                    DeltaT const* row_delta,
+                    column_device_view::const_iterator<OrderbyT> begin,
+                    column_device_view::const_iterator<OrderbyT> end)
+      : groups{groups}, row_delta{row_delta}, begin{begin}, end{end}
+    {
+    }
+    Grouping groups;  ///< Group information to determine bounds on current row's window
+    static_assert(cuda::std::is_same_v<Grouping, ungrouped> ||
+                    cuda::std::is_same_v<Grouping, grouped> ||
+                    cuda::std::is_same_v<Grouping, ungrouped_with_nulls> ||
+                    cuda::std::is_same_v<Grouping, grouped_with_nulls>,
+                  "Invalid grouping descriptor");
+    DeltaT const* row_delta;  ///< Delta from current row that defines the interval endpoint. This
+                              ///< pointer is null for UNBOUNDED and CURRENT_ROW windows.
+    column_device_view::const_iterator<OrderbyT>
+      begin;                                           ///< Iterator to beginning of orderby column
+    column_device_view::const_iterator<OrderbyT> end;  ///< Iterator to end of orderby column
+
+    /**
+     * @brief Compute the offset to the end of the window.
+     *
+     * @param i The current row index.
+     * @return Offset to the current row's window endpoint.
+     */
+    [[nodiscard]] __device__ size_type operator()(size_type i) const
+    {
+      using Comp      = comparator_t<OrderbyT, WindowTag, Order>;
+      auto const info = groups.row_info(i);
```

Comment:
I might have missed where `info` is used. We seem also to be unpacking the result of `groups.row_info(i)` on the next line.  

Did we intend to remove the `info` line?

Reply:
Yes, should be removed, thanks.